### ISSUE TITLE
Pure-GNN v3.1 v2 checkpoint roundtrip diagnosis and Kaggle synthetic preflight

### ISSUE THREAD

**Issue by Irthn1311** [Irthn1311/FER2013_Graph#73]
## Scope

Investigate the immutable Pure-GNN v3.1 scientific-screen v1 technical failure that occurred during the Kaggle pytest gate before any FER Train/Val/Test access or scientific training.

### Immutable v1 reference

- Branch: `research/pure-gnn-v31`
- V1 tag: `pure-gnn-v31-scientific-screen-v1`
- V1 peeled commit: `f79d748f154d7aea21362e29d1ead86c4e552599`
- V1 classification: **TECHNICAL FAILURE DURING PRE-DATA PYTEST GATE**
- V1 must never be moved, deleted, force-updated, or reused.

### Observed Kaggle failure

Runtime:
- Python 3.12.13
- TensorFlow 2.20.0
- 2 GPUs

Pytest:
- 140 passed
- 3 failed

Failures:
- `test_checkpoint_weights_roundtrip[G0]`
- `test_checkpoint_weights_roundtrip[G0.5]`
- `test_checkpoint_weights_roundtrip[G1]`

Observed restored-logit differences were approximately O(1e-6), while the existing test requires `max_abs(logits_before - logits_after) <= 1e-7`.

The notebook stopped at pytest, so:
- real Train was not accessed;
- real Val was not accessed;
- PrivateTest was not accessed;
- no G0/G0.5/G1 scientific training occurred.

## Primary objective

Determine whether the v1 failure is:

1. **CHECKPOINT_SERIALIZATION_CORRUPTION**, or
2. **NUMERICAL_FORWARD_DRIFT_WITH_EXACT_WEIGHTS**.

Do not assume the second. Prove exact checkpoint-state preservation first.

## Required checkpoint diagnostic

For each of `G0`, `G0.5`, and `G1`, using synthetic input only:

1. Set deterministic Python/NumPy/TensorFlow seeds.
2. Instantiate/build original model.
3. Run the exact same fixed input multiple times and record same-model repeat drift.
4. Snapshot all model variables with stable identity/path, dtype, shape, and exact array value.
5. Save using the existing scientific `.weights.h5` format.
6. Instantiate/build a fresh same-condition model.
7. Load the saved weights.
8. Snapshot all restored variables.
9. Require identical variable set/count/path mapping, shape, and dtype.
10. Compare every corresponding tensor with exact equality (`np.testing.assert_array_equal`).
11. Record maximum absolute parameter difference.
12. Run restored model multiple times on the same fixed input.
13. Record original-repeat, restored-repeat, and original-vs-restored logit differences.
14. Record checkpoint file SHA256 and size.

### Decision rule

If any parameter tensor differs after save/load:
- classify as `CHECKPOINT_SERIALIZATION_CORRUPTION`;
- stop;
- do not relax tolerance;
- do not create v2;
- do not access FER data.

If every parameter tensor restores exactly (`max_parameter_difference == 0.0`) but logits differ only at floating-point micro-scale, classify as `NUMERICAL_FORWARD_DRIFT_WITH_EXACT_WEIGHTS`.

Only in that second case may the test contract be split into:

- serialization contract: exact restored tensor equality, zero tolerance;
- functional numerical contract: `max_abs(logits_before - logits_after) <= 1e-5`.

The `1e-5` functional tolerance must not be increased.

## V2 candidate patch constraints

Only if exact weight restoration passes for all three conditions:

- Refactor `research/pure_gnn_v31/tests/test_scientific_checkpointing.py` to verify exact parameter restoration plus `<=1e-5` functional logit equivalence.
- Do **not** modify checkpoint implementation, trainer, architecture, optimizer, scheduler, loss, augmentation, seed, batch size, or scientific recipe.
- Update the authorized notebook/tag contract from `pure-gnn-v31-scientific-screen-v1` to `pure-gnn-v31-scientific-screen-v2`.
- Update authorization-state tests to require the exact v2 tag string; do not weaken them to arbitrary tags.

Expected narrow changed paths:
- `research/pure_gnn_v31/tests/test_scientific_checkpointing.py`
- `notebooks/pure-gnn-v31-scientific-screen-kaggle.ipynb`
- `research/pure_gnn_v31/tests/test_scientific_notebook_contract.py`
- optionally one dedicated synthetic checkpoint diagnostic test.

## Local validation

Run:

```bash
python -m pytest research/pure_gnn_v31/tests -q
git diff --check
```

Compile all notebook code cells.

Commit candidate as:

`Refine Pure-GNN checkpoint roundtrip contract for v2`

Push normally to `research/pure-gnn-v31` and record the exact `V2_CANDIDATE_SHA`.

Do **not** create the v2 tag yet.

## Kaggle synthetic technical preflight

Using the exact `V2_CANDIDATE_SHA`, launch a Kaggle technical preflight with synthetic data only:

- no FER dataset attached;
- detached checkout of exact candidate SHA;
- direct package import from `research/pure_gnn_v31/src`;
- print Python, TensorFlow, GPU list, exact HEAD;
- run at minimum `test_scientific_checkpointing.py`, preferably the full Pure-GNN test suite;
- run the explicit checkpoint diagnostic for G0/G0.5/G1;
- do not execute the scientific screen.

Target environment should match the v1 failure runtime as closely as available (TensorFlow 2.20-class Kaggle runtime, T4 GPU if available).

### Kaggle preflight pass criteria

For G0/G0.5/G1:
- exact parameter equality: PASS;
- max parameter difference: `0.0`;
- functional restored-logit error: `<=1e-5`;
- full pytest: PASS;
- FER data access: none.

If any criterion fails, stop and do not create v2 tag.

## Governance

Must remain unchanged:
- architecture;
- trainer/scientific training logic;
- checkpoint implementation;
- optimizer/scheduler/loss/augmentation;
- scientific recipe;
- Train/Val/Test protocol.

No real FER Train/Val/Test access in this issue.
No scientific training in this issue.
No v2 tag creation in this issue.
Never move v1.

## Expected final status

If exact weights and Kaggle synthetic preflight pass, report:

`PURE_GNN_V31_V2_CANDIDATE_READY_FOR_REVIEW`

If parameter restoration fails:

`PURE_GNN_V31_CHECKPOINT_SERIALIZATION_FAILURE`

If Kaggle technical preflight fails for another reason:

`PURE_GNN_V31_V2_TECHNICAL_PREFLIGHT_FAILURE`

**Comment by Irthn1311** [comment]
EXECUTE NOW, end-to-end, without adding another staging/preflight layer. Follow the updated Issue #73 exactly: keep architecture/trainer/recipe/checkpoint implementation unchanged; fix the checkpoint roundtrip contract only if restored parameters are exactly equal; update notebook/tag contract to v2; run the full test suite; create immutable `pure-gnn-v31-scientific-screen-v2`; submit the real Kaggle notebook; wait for completion; verify artifacts and report G0/G0.5/G1 seed-42 results. Stop only on the explicit Issue stop conditions. Do not touch PrivateTest and do not move v1.

**Comment by Irthn1311** [comment]
EXECUTE V3 NOW. Apply only the authorized minimal notebook role fix (`val` -> governance role `validation`), update the exact authorized tag contract to `pure-gnn-v31-scientific-screen-v3`, run the required tests, create immutable v3, then submit the real Kaggle seed-42 G0/G0.5/G1 scientific screen immediately. Do not add another preflight/staging layer. Never move v1 or v2. Do not touch PrivateTest.
